In [22]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Bhati\OneDrive\Desktop\CF\Data Analytics Immersion\Achievement 4\11-Apr-2023 Instacart Basket Analysis'

In [3]:
# importing the new dataset of customers
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

# Data Wrangling

In [4]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# renaming columns
df_cust.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'Number_of_dependents', 'fam_status' : 'family_status'}, inplace = True)

In [6]:
df_cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number_of_dependents family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [7]:
df_cust.shape

(206209, 10)

In [9]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   first_name            194950 non-null  object
 2   last_name             206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   Number_of_dependents  206209 non-null  int64 
 8   family_status         206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


Some first names are missing in the dataset

In [10]:
# changing user_id data type from int64 to string
df_cust['user_id'] = df_cust['user_id'].astype('str')

In [11]:
df_cust.describe()

age  Number_of_dependents         income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

In [12]:
# Checking for mixed data type
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)
    print (len(df_cust[weird]))

first_name
11259


There are 11259 non-uniform values in the first_name column.

In [14]:
# Checking for duplicates
df_duplicates = df_cust[df_cust.duplicated()]

No duplicates

In [17]:
#importing final aggregated file
ords_prods_agg = pd.read_pickle(os.path.join(path,'02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

In [18]:
ords_prods_agg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   product_id              int64  
 1   product_name            object 
 2   aisle_id                int64  
 3   department_id           int64  
 4   prices                  float64
 5   order_id                int64  
 6   user_id                 int64  
 7   order_number            int64  
 8   orders_day_of_week      int64  
 9   order_hour_of_day       int64  
 10  days_since_prior_order  float64
 11  add_to_cart_order       int64  
 12  reordered               int64  
 13  price_range_loc         object 
 14  busiest day             object 
 15  busiest days            object 
 16  busiest_period_of_day   object 
 17  max_order               int64  
 18  loyalty_flag            object 
 19  avg_price               float64
 20  spending_flag           object 
 21  median_freq             float

In [19]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  object
 1   first_name            194950 non-null  object
 2   last_name             206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   Number_of_dependents  206209 non-null  int64 
 8   family_status         206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 15.7+ MB


In [20]:
# changing user_id data type from object to int64
df_cust['user_id'] = df_cust['user_id'].astype('int64')

In [21]:
# merging dataframes

ords_prods_cust = ords_prods_agg.merge(df_cust, on = 'user_id')

In [23]:
ords_prods_cust.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2         907        Premium Sliced Bacon       106             12    20.0   
3         907        Premium Sliced Bacon       106             12    20.0   
4        1000                    Apricots        18             10    12.9   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2   3160996      138             1                   5                 13   
3   2254091      138            10                   5                 14   
4    505689      138             9                   6                 12   

   ...    order_freq_flag  first_name  last_name gender      state age  \
0  ...  Frequent customer     Charles        Cox   Male  Minnesota  81   
1  ...  Frequent customer     Charles        Cox   Male  Minnesota  81   
2  ...  Frequent customer     Charles        Cox   Male  Minnesota  81   
3  ...  Frequent customer     Charles        Cox   Male  Minnesota  81   
4  ...  Frequent customer     Charles        Cox   Male  Minnesota  81   

  date_joined  Number_of_dependents family_status  income  
0    8/1/2019                     1       married   49620  
1    8/1/2019                     1       married   49620  
2    8/1/2019                     1       married   49620  
3    8/1/2019                     1       married   49620  
4    8/1/2019                     1       married   49620  

[5 rows x 32 columns]

In [25]:
ords_prods_cust.shape

(32404859, 32)

In [26]:
ords_prods_cust.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 32 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   product_id              int64  
 1   product_name            object 
 2   aisle_id                int64  
 3   department_id           int64  
 4   prices                  float64
 5   order_id                int64  
 6   user_id                 int64  
 7   order_number            int64  
 8   orders_day_of_week      int64  
 9   order_hour_of_day       int64  
 10  days_since_prior_order  float64
 11  add_to_cart_order       int64  
 12  reordered               int64  
 13  price_range_loc         object 
 14  busiest day             object 
 15  busiest days            object 
 16  busiest_period_of_day   object 
 17  max_order               int64  
 18  loyalty_flag            object 
 19  avg_price               float64
 20  spending_flag           object 
 21  median_freq             float

# Exporting the combined file

In [27]:
ords_prods_cust.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customer.pkl'))